### PR TITLE
[ACM-10004] Added template override annotations

### PR DESCRIPTION
**Pull request by dislbenn**
# Description

Adding a method to allow template overrides within the MCH operator. These overrides will allow users to define a `TEMPLATE_OVERRIDE_*` env variable or template override `configmap` that can be selected by annotation in the MCH namespace.

## Related Issue

https://issues.redhat.com/browse/ACM-10004

## Changes Made

Added template overrides logic to MCH operator.

## Screenshots (if applicable)

Add screenshots or GIFs that demonstrate the changes visually, if relevant.

## Checklist

- [ ] I have tested the changes locally and they are functioning as expected.
- [x] I have updated the documentation (if necessary) to reflect the changes.
- [x] I have added/updated relevant unit tests (if applicable).
- [x] I have ensured that my code follows the project's coding standards.
- [x] I have checked for any potential security issues and addressed them.
- [x] I have added necessary comments to the code, especially in complex or unclear sections.
- [x] I have rebased my branch on top of the latest main/master branch.

## Additional Notes

Add any additional notes, context, or information that might be helpful for reviewers.

## Reviewers

Tag the appropriate reviewers who should review this pull request. To add reviewers, please add the following line: `/cc @reviewer1 @reviewer2`

## Definition of Done

- [ ] Code is reviewed.
- [ ] Code is tested.
- [ ] Documentation is updated.
- [ ] All checks and tests pass.
- [ ] Approved by at least one reviewer.
- [ ] Merged into the main/master branch.

### CODE REVIEW

File: controllers/common.go [hunk 1]
```diff
@@ -324,13 +327,65 @@ func (r *MultiClusterHubReconciler) OverrideImagesFromConfigmap(imageOverrides m
...
+		}
+	}
+
+	for key, value := range config.TemplateOverrides {
+		// Convert value to string if necessary
+		strValue, err := convertToString(value)
+		if err != nil {
+			return nil, fmt.Errorf("error converting value for key %s: %w", key, err)
+		}
+		templateOverrides[key] = strValue
```
Comment (joeg-pro) [inline controllers/common.go:368]: If I'm following along correctly, in _controller.go, we first read in template overrides from env vars, and then if there is an template-overrides-cm annotation we call this function passing the templateOverides map we have after the load-from-env-var stuff.  Here, in this function we iterate over the keys in the json obj from the config map, and for each key was save a key/value entry into the templateOVerrides map.

This achieve a merging of the two sources,  with the effect that if a given key appears in both sources, the value from the configmap key/value entry wins.  Its a close call either way, but I would prefer to see this implemented such that if  given key appears in both sources, the value from the environment variable wins.    My rationale is that such an approach would let us have some "permanent overrides" that are established by creating a config map and making the MCH point to it, and then you could have some "temporary overrides of those overrides" by setting env vars (via OLM subscription spec.config.env).

If you agree, one way to do that would be to check that the key is not already in the map here, before setting the map entry.

File: pkg/imageoverrides/templateoverrides.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	return templateOverrides
+}
+
+// parseEnvVar parses the environment variable and extracts key and value
+func parseEnvVar(envVar string) (key string, value string) {
+	pair := strings.SplitN(envVar, "=", 2)
+	if len(pair) == 2 && strings.HasPrefix(pair[0], TemplateOverridePrefix) {
+		key = strings.ToLower(strings.TrimPrefix(pair[0], TemplateOverridePrefix))
```
Comment (joeg-pro) [inline pkg/imageoverrides/templateoverrides.go:40]: I just realized I forgot to mention that we should fold the key-name extracted from the env var name to lower case (for consistency with imageOverrides and other maybe subtle reasons) and came here to make a comment and find its already being done.  Great!

**Review by joeg-pro — CHANGES_REQUESTED** [review]
I have one change I'd like to see (rre the precedence order) and some other sugs but this looks almost good to go.

File: pkg/imageoverrides/templateoverrides.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2020 Red Hat, Inc.
+// Copyright Contributors to the Open Cluster Management project
+
+package imageoverrides
```
[the remark this answers is not in the document]
Comment (dislbenn) [reply pkg/imageoverrides/templateoverrides.go:4]: Yes, we can rename it to a more general package name like `overrides`.

**Review by joeg-pro — APPROVED** [review]
Thanks for the updates.

**Pull request merged**